File: CSharp/Legacy89DiskKit/Fdc/Application/FdcAccessService.cs
using Legacy89DiskKit.Domain.Fdc.Interface;
using Legacy89DiskKit.Domain.Fdc.Model;
using Legacy89DiskKit.Domain.Timing.Interface;

namespace Legacy89DiskKit.Fdc.Application;

public class FdcAccessService
{
    private readonly IFdcController _controller;
    private readonly IControllerClock? _clock;

    public FdcAccessService(IFdcController controller, IControllerClock? clock = null)
    {
        _controller = controller ?? throw new ArgumentNullException(nameof(controller));
        _clock = clock;
    }

    public bool SupportsTimingAdvance => _clock is not null;

    public void Reset()
    {
        _controller.Reset();
    }

    public void WriteRegister(FdcRegister register, byte value)
    {
        _controller.WriteRegister(register, value);
    }

    public byte ReadRegister(FdcRegister register)
    {
        return _controller.ReadRegister(register);
    }

    public FdcVisibleState GetVisibleState()
    {
        return _controller.GetVisibleState();
    }

    public void Advance(TimeSpan delta)
    {
        if (_clock is null)
        {
            throw new InvalidOperationException("Timing advance is not available without a controller clock.");
        }

        _clock.Advance(delta);

        if (_controller is ITimedFdcController timedController)
        {
            timedController.Advance(delta);
        }
    }
}

File: Cpp/Legacy89DiskKit.Cpp/src/application/fdc_access_service.cpp
#include "legacy89diskkit/cpp/application/fdc_access_service.hpp"
#include <stdexcept>

namespace legacy89diskkit::cpp::application
{
FdcAccessService::FdcAccessService(
    std::shared_ptr<FdcController> controller,
    std::shared_ptr<ControllerClock> clock)
    : controller_(std::move(controller)), clock_(std::move(clock))
{
    if (!controller_)
    {
        throw std::invalid_argument("controller cannot be null");
    }
}

bool FdcAccessService::SupportsTimingAdvance() const
{
    return clock_ != nullptr;
}

void FdcAccessService::Reset()
{
    controller_->Reset();
}

void FdcAccessService::WriteRegister(FdcRegister reg, std::uint8_t value)
{
    controller_->WriteRegister(reg, value);
}

std::uint8_t FdcAccessService::ReadRegister(FdcRegister reg)
{
    return controller_->ReadRegister(reg);
}

FdcVisibleState FdcAccessService::GetVisibleState() const
{
    return controller_->GetVisibleState();
}

void FdcAccessService::Advance(std::chrono::nanoseconds delta)
{
    if (!clock_)
    {
        throw std::logic_error("Timing advance is not available without a controller clock.");
    }

    // Update the master clock for the controller environment
    clock_->Advance(delta);

    // Notify the controller to update its internal state based on the same delta.
    // Note: If the controller and clock share the same internal state object, 
    // it is up to the implementation to avoid double-advancement.
    // In our default wiring, they are separate observers of the timeline.
    auto timed_controller = std::dynamic_pointer_cast<TimedFdcController>(controller_);
    if (timed_controller)
    {
        timed_controller->Advance(delta);
    }
}
} // namespace legacy89diskkit::cpp::application
